Support direct editing of the intelRdt config

marquiz opened the pull request:
This PR adds support for directly adjusting the [`intelRdt`](https://github.com/opencontainers/runtime-spec/blob/main/config-linux.md#intelrdt) object of the container config.

Previously, we have the concept of RDT class, which provides indirect manipulation of the `intelRdt.closID` field. There the higher level runtime (containerd, cri-o) registers a resolver function which the NRI adaptation uses to translate the provided "RDT class" to actual closID (which is basically the resctrl group in the Linux resctrl filesystem).

The API extensions in this PR provides direct control of the relevant fields of the `intelRdt` object. The PR also includes a sample plugin to test/demonstrate the functionality (including deployment files and CI integration to build/publish images)

**REFS:** 
- https://github.com/opencontainers/runtime-spec/pull/1230
- https://github.com/opencontainers/runtime-spec/pull/1287

**NOTES:**
- ~~Depends/stacks on #212. That could be merged separately~~
- The "pkg/adaptation: support new RDT fields" commit borrows some snippets from #118.
- Known runtime implementations are runc (upcoming v1.4) and crun (v1.24+)
  - `closID`: supported by runc and crun
  - `schemata`: supported by runc  (upcoming v1.4) and crun (v1.22+)
    - runc PR: https://github.com/opencontainers/runc/pull/4830
    - crun PR: https://github.com/containers/crun/pull/1748
   - `enableMonitoring`  supported by runc  (upcoming v1.4) but not by crun (v1.24+)
     - runc PR: https://github.com/opencontainers/runc/pull/4832)
     - crun PR: https://github.com/containers/crun/pull/1866

**OPEN QUESTION/TODO:**

In this PR there is no mechanism for removing the `intelRdt` object from the config. That is probably a desirable feature.

```release-note
Support direct adjustment of the intelRdt container configuration
```

marquiz committed:
api: add rdt

Add new message for directly managing the linux.intelRdt object of the
runtime-spec.

NOTE: the old RdtClass field resides under LinuxResources message but
the IntelRdt message is sepate from LinuxResources. This is to maintain
parity with the OCI runtime-spec. It is also much more logical because
the LinuxResources message is used in many places (like pod overhead,
update resources etc) where the RDT fields don't make sense or cannot be
applied. Also note that the Rdt object cannot be modified on container
update (missing from UpdateContainerRequest) as this functionality is
missing from the runtime spec and runtimes (runc/crun).

Signed-off-by: Markus Lehtonen <markus.lehtonen@intel.com>

marquiz committed:
pkg/runtime-tools/generate: support new RDT fields

Add support for direct manipulation of the linuxRdt object in the
container config.

Signed-off-by: Markus Lehtonen <markus.lehtonen@intel.com>

marquiz committed:
pkg/adaptation: support new RDT fields

Add support for direct manipulation of the intelRdt object of the
container config.

Co-authored-by: Krisztian Litkey <krisztian.litkey@intel.com>
Signed-off-by: Markus Lehtonen <markus.lehtonen@intel.com>

marquiz committed:
plugins: add sample rdt plugin

Signed-off-by: Markus Lehtonen <markus.lehtonen@intel.com>

klihub commented on pkg/api/adjustment.go:
```diff
@@ -378,3 +396,9 @@ func (a *ContainerAdjustment) initLinuxResourcesUnified() {
 		a.Linux.Resources.Unified = make(map[string]string)
 	}
 }
+func (a *ContainerAdjustment) initLinuxRdt() {
```
Just a note: #118 has a bunch of queued similar changes, updating the result-handling code so that there is no need to always a priory create a fully constructed but empty adjustment. It would be nice to get that finally in, so I'll rebase it that on the latest main/HEAD then update with any missing init*() bits necessary.

marquiz replied:
:+1: I'll review when that's updated. And then rebase when that's merged

marquiz commented on pkg/api/update.go:
```diff
@@ -16,7 +16,7 @@
 package api
 
-//nolint
+// nolint
```
Yes. I didn't even notice this change :roll_eyes:, was automatic by gofmt. I now changed this to more appropriate format.

marquiz commented:
I just recalled that there's actually one aspect/feature missing from the new extensions: ability to remove the `linux.intelRdt` from the container config (set it to nil/null). Suggestions for how to implement this are warmly welcome.

I couldn't come up with any nice solutions, I played with two alternatives but didn't commit either of them:
1. Have a separate field (like `drop`) in the NRI API
2. Have a well documented, otherwise invalid value for the `closID` field in the NRI API (e.g. `-/`) that causes the config to be dropped

giuseppe reviewed (approved):
LGTM

klihub commented:
> I just recalled that there's actually one aspect/feature missing from the new extensions: ability to remove the `linux.intelRdt` from the container config (set it to nil/null). Suggestions for how to implement this are warmly welcome.
> 
> I couldn't come up with any nice solutions, I played with two alternatives but didn't commit either of them:
> 
> 1. Have a separate field (like `drop`) in the NRI API
> 2. Have a well documented, otherwise invalid value for the `closID` field in the NRI API (e.g. `-/`) that causes the config to be dropped

@marquiz I'd vote with a mild bias for 2 above (because it's how we handle other similar cases where removal is supported). But with the addition that on the wire-protocol it's convention we use, but in the user-facing API we provide explicit functions for clearing any LinuxRdt,  with something along the lines of this:

```golang
func (a *ContainerAdjustment) RemoveLinuxRDT() {
    a.initLinux()
    a.Linux.Rdt = &LinuxRdt{
        ClosId = MarkForRemoval("") // or MarkForRemoval("/") if we want to keep
    }
}
```

marquiz commented:
@klihub thanks for the feedback. The MarkForRemoval thingie was a good hint, I didn't have that in my early prototype. I'll cook up something along the lines you suggested and then we can mull over that.

marquiz committed:
Implement removal of RDT

Makes it possible to remove/override the linux.intelRdt object from the
container configuration. Extend the API by adding new 'Remove' field to
the LinuxRdt message which is used as a marker to fully delete/override
the IntelRdt configuration. This patch also updates the adaptation
and runtime-tools correspondingly.

Signed-off-by: Markus Lehtonen <markus.lehtonen@intel.com>

klihub commented on go.mod:
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/onsi/ginkgo/v2 v2.19.1
 	github.com/onsi/gomega v1.34.0
-	github.com/opencontainers/runtime-spec v1.1.0
+	github.com/opencontainers/runtime-spec v1.2.2-0.20250818071321-383cadbf08c0
```
@marquiz @mikebrow @chrishenzie In principle there is one way how we could get this partially in already now, to avoid having to keep rebasing this with a guaranteed conflict whenever we add something to LinuxContainer or LinuxContainerAdjustment.

It would be to split this up to 2 stacked PRs:
- PR#1: The bulk of this PR, without the OCI Spec bump + a single change to Generator.Adjust(): return an error if adjust.GetLinux().GetRdt() != nil
- draft PR#2: All the commits in PR#1 + the OCI Spec bump (eventually to a tagged version) + a Generator.Adjust() updated to do RDT adjustment if requested

Then we could merge #PR1 (and rebase PR#2 when that happens), and keep the OCI Spec-dependent bits pending until the necessary bits get tagged. We could apply the same to #157 and #166...

Maybe not worth the effort though, and we can just mark this a draft until we get a tagged OCI Spec.

marquiz commented:
> I'd vote with a mild bias for 2 above (because it's how we handle other similar cases where removal is supported). But with the addition that on the wire-protocol this is the convention we use, but in the user-facing API we provide explicit functions for clearing any LinuxRdt,

I ended up with adding a separate field. I implemented both variants and the "prefix closid with `-/`" variant was an ugly fragile mess tbh. So much cleaner to just have a separate field for this purpose. For envs and other guys it's a bit different when we're updating a map or slice and you selectively remove some items. But here we're managing the whole object. (if interested, I have the still lying around at https://github.com/marquiz/nri/commits/devel/rdt-remove-2/)

klihub commented:
> @klihub thanks for the feedback. The MarkForRemoval thingie was a good hint, I didn't have that in my early prototype. I'll cook up something along the lines you suggested and then we can mull over that.

@marquiz Sorry about the long sidetrack in our offline discussion about Remove*(). I was totally wrong and in the wrong mental model associated with most of the existing bits that support remove. We have a single RDT field (not a slice) so you were absolutely right that we can't handle removal with a MarkForRemoval inline notation, because then it would not be possible to indicate a removal prior to an addition in order to avoiding conflict with an earlier plugin. I don't know what the hell I was thinking... or I think I wasn't actually thinking at all.

marquiz commented:
Rebased, marked ready-for-review. Stacked on top of #212 
@klihub @mikebrow @chrishenzie

klihub commented:
As a purely theoretical question related to your notes in the commit log of  'api: add rdt'.

In the (I am adraid highly unlikely) situation where runc/crun/OCI compatible runtimes would gain the ability to update RDT for a running container, what would be the best way for us to model/handle that. Would we add a corresponding `&LinuxRdt{}` to `LinuxContainerUpdate{}` ?

marquiz commented:
> In the (I am adraid highly unlikely) situation where runc/crun/OCI compatible runtimes would gain the ability to update RDT for a running container, what would be the best way for us to model/handle that. Would we add a corresponding `&LinuxRdt{}` to `LinuxContainerUpdate{}`

Yes, I think that would likely be the place (with the usual reservations of predicting the future, how this would be implemented in runc).

marquiz commented:
@klihub @mikebrow @chrishenzie it would be nice to get #118 in before this. I'll then rebase

marquiz commented:
> @marquiz We could also do this the other way around. I think we're more eager to get this in. And I wouldn't mind at all that merge order forcing me to do another rebase of #118,

OK, thanks @klihub. I now rebased this PR.

I added you as Co-author in the `pkg/adaptation: support new RDT fields` patch where I borrowed some snippets from #118.

chrishenzie commented on pkg/adaptation/result.go:
```diff
@@ -499,6 +503,42 @@ func (r *result) adjustSysctl(sysctl map[string]string, plugin string) error {
+
+	if rdt.GetRemove() {
+		r.owners.ClearRdt(id, plugin)
+		r.reply.adjust.Linux.Rdt = &LinuxRdt{}
```
Does `Remove` need to be set on this as well in order for it to be removed from the OCI spec?

marquiz replied:
Thanks for sharp-eyed review @chrishenzie. Indeed, it needs to be set so that if the original oci spec (before any nri plugins kicked in) contains RDT configuration, the final apply/adjust will then remove it. Fixed.

klihub commented on pkg/api/owners.go:
```diff
@@ -773,6 +819,14 @@ func (f *FieldOwners) RdtClassOwner() (string, bool) {
 	return f.simpleOwner(Field_RdtClass.Key())
 }
 
```
I'd say we do want it as it would be very useful for finer-grained validation if/once we manage to kick an eventually accepted version of that (with reliable plugin identification/authentication) forward. But I think it's already there: https://github.com/containerd/nri/blob/082b99b5b60a6273b0845b83853cfeb7e507a05a/pkg/api/owners.go#L339

chrishenzie replied:
Oops, thanks for pointing that out

klihub commented:
@mikebrow Mike, how about this one ? I labelled this v.next originally, before it got reviewed. But this is pretty self-contained and should be safe to merge, so I'd be okay with pushing this in already for v0.11.0.

mikebrow reviewed (approved):
LTGM
Well done!
